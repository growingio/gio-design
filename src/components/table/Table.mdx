import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Table from './index';

# Table 列表

<Subtitle>通用列表。</Subtitle>

## 参数说明

<ArgsTable of={Table} />

## 代码演示

### 基础

合适的数据展示方式可以帮助用户快速地定位和浏览数据/信息，以及更高效得协同工作。列表常和排序、搜索、筛选、分页等其他界面元素一起协同，适用于信息的收集和展示、归纳整理、以及操作结构化数据/信息。它结构简单，分隔归纳明确，使信息之间更易于阅读，大大提升了用户对信息的接收效率和理解程度。

- 列表标题：标题是对列表信息内容的整体概括，以便用户对列表内通有整体认知。
- 表头：一般指列标签，这里也指首列行标签，是对所属列/行的信息描述。
- 表体：列表主体内容，具体信息内容的填充区域。

[Example](/?path=/story/functional-components-table--base)

### 表头多样式

在使用列表的过程中，表头常常与其他辅助性控件，如提示、排序、过滤等对快速获取/阅读列表数据/信息提供帮助的辅助性控件配合使用。
【常用辅助控件】：一般对当前表头所属列的内容进行解释，便于用户理解。

- 提示：一般对当前表头所属列的内容进行解释，便于用户理解。
- 排序：可以对当前列进行升序/降序排序。
- 过滤：可以快速定位当前列的某个元素。

【组合规则】：我们常常会遇到同一个表头使用多个辅助性控件的情况，使用/排列规则如下：

- 提示：始终跟随表头文案，居其右侧排列。
- 排序：始终在表头所属表格居右显示；当过滤等其他操作设置也居右显示时，排序控件位置不变，其他操作设置随其左侧显示。
- 过滤：正常情况下，在所属表格居右显示；当「排序」也居右显示时，过滤居其左。
- 不建议一个表头中超过 2 个以上的辅助性控件。多控件的出现，反而会影响用户阅读信息的效率。

[Example](/?path=/story/functional-components-table--table-header)

### 多行表头

[Example](/?path=/story/functional-components-table--multi-line)

### 滚动

[Example](/?path=/story/functional-components-table--table-scroll)

### 无结果

[Example](/?path=/story/functional-components-table--table-empty)

### 列表加载

[Example](/?path=/story/functional-components-table--table-loading)

### 分页

[Example](/?path=/story/functional-components-table--table-pagination)
