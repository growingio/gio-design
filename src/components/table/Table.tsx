import React, { useMemo, forwardRef, createContext } from 'react';
import RcTable from 'rc-table';
import classNames from 'classnames';
import { cloneDeep, isUndefined, get, has, set } from 'lodash';
import { compose } from 'lodash/fp';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import useMergeRef from '../../utils/hooks/useMergeRef';
import useSorter from './hook/useSorter';
import useFilter from './hook/useFilter';
import usePagination from './hook/usePagination';
import useSelection from './hook/useSelection';
import useEllipsisTooltip from './hook/useEllipsisTooltip';
import Title from './Title';
import { TableProps, ColumnsType } from './interface';
import Empty from './Empty';
import { translateInnerColumns } from './utils';
import Loading from '../loading';
import useDebounceLoading from '../../utils/hooks/useDebounceLoading';
import useHackOnRow from './hook/useHackOnRow';

interface TableContextType {
  tableRef: null | React.MutableRefObject<HTMLDivElement>;
}
export const TableContext = createContext({ tableRef: null } as TableContextType);

const Table = <RecordType,>(
  props: TableProps<RecordType>,
  ref: React.MutableRefObject<HTMLDivElement>
): React.ReactElement => {
  const {
    prefixCls: customizePrefixCls,
    title,
    columns = [],
    dataSource = [],
    pagination = {},
    rowSelection,
    showIndex = false,
    emptyText = null,
    onChange,
    showHover = true,
    rowKey,
    loading = false,
    onRow,
    hackRowEvent = false,
    className,
    style,
    ...rest
  } = props;
  const mergedRef = useMergeRef(ref);
  const prefixCls = usePrefixCls('table', customizePrefixCls);
  const debounceLoading = useDebounceLoading(loading, 1000);
  const onHackRow = useHackOnRow(onRow, hackRowEvent);
  const innerColumns = useMemo(() => translateInnerColumns(columns), [columns]);
  const [activeSorterStates, updateSorterStates, sortedData] = useSorter(innerColumns, dataSource);
  const [activeFilterStates, updateFilterStates, filtedData] = useFilter(innerColumns, sortedData);
  const [
    transformShowIndexPipeline,
    activePaginationedState,
    paginationedData,
    PaginationComponent,
    resetPagination,
  ] = usePagination(filtedData, pagination, showIndex);

  const [transformSelectionPipeline] = useSelection(paginationedData, rowSelection, {
    rowKey,
  });
  const [transformEllipsisTooltipPipeline] = useEllipsisTooltip();

  const onTriggerStateUpdate = (reset = false, paginationState = activePaginationedState): void => {
    if (reset) {
      resetPagination();
    }
    // 通过 activePaginationedState 拿不到最新的状态
    onChange?.(paginationState, activeSorterStates, activeFilterStates);
  };

  const renderTitle = (_columns: ColumnsType<RecordType>): ColumnsType<RecordType> =>
    cloneDeep(_columns).map((column) => {
      const sortState = activeSorterStates.find(({ key }) => key === column.key);
      const filterState = activeFilterStates.find(({ key }) => key === column.key);
      if (sortState || filterState || !isUndefined(column.info)) {
        const oldColumn = cloneDeep(column);
        set(
          column,
          'title',
          <Title
            sorterState={sortState}
            filterState={filterState}
            prefixCls={prefixCls}
            column={oldColumn}
            updateSorterStates={updateSorterStates}
            updateFilterStates={updateFilterStates}
            onTriggerStateUpdate={onTriggerStateUpdate}
          />
        );
      }
      if (has(column, 'children')) {
        set(column, 'children', renderTitle(get(column, 'children')));
      }
      return column;
    });

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const transformColumns = useMemo(() => renderTitle(innerColumns), [
    activeSorterStates,
    activeFilterStates,
    innerColumns,
    prefixCls,
  ]);

  const composedColumns = compose(
    transformEllipsisTooltipPipeline,
    transformSelectionPipeline,
    transformShowIndexPipeline
  )(transformColumns);

  const emptyElement = emptyText || (
    <div className={`${prefixCls}-empty`}>
      <Empty />
      <p>无搜索结果</p>
    </div>
  );

  return (
    <TableContext.Provider value={{ tableRef: mergedRef }}>
      <div
        className={classNames(`${prefixCls}-wrapper`, className, {
          [`${prefixCls}-showHover`]: showHover,
        })}
        style={style}
        ref={mergedRef}
      >
        <Loading loading={debounceLoading}>
          <RcTable
            title={title ? () => title : undefined}
            prefixCls={prefixCls}
            columns={composedColumns}
            data={paginationedData}
            emptyText={emptyElement}
            rowKey={rowKey}
            onRow={onHackRow}
            {...rest}
          />
          <PaginationComponent onTriggerStateUpdate={onTriggerStateUpdate} />
        </Loading>
      </div>
    </TableContext.Provider>
  );
};

export default forwardRef(Table);
