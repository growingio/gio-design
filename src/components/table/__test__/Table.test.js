import React from 'react';
import { mount } from 'enzyme';
import { render } from '@testing-library/react';
import { act } from 'react-test-renderer';
import Table from '../index';

const dataSource = [
  {
    key: '1',
    name: '胡彦斌',
    age: 32,
    weight: 70,
  },
  {
    key: '2',
    name: '胡彦祖',
    age: 42,
    weight: 60,
  },
  {
    key: '3',
    name: '航航',
    age: 18,
    weight: 70,
  },
  {
    key: '4',
    name: '屁屁',
    age: 22,
    weight: 60,
  },
];

const columns = [
  {
    title: '姓名',
    dataIndex: 'name',
    key: 'name',
    info: '这里是用户的姓名',
    sorter: (a, b) => a.name.length - b.name.length,
    filters: ['名字俩字', '名字仨字'],
    onFilter: (value, record) => {
      if (value === '名字俩字') {
        return record.name.length === 2;
      }
      if (value === '名字仨字') {
        return record.name.length === 3;
      }
      return false;
    },
  },
  {
    title: '年龄',
    dataIndex: 'age',
    key: 'age',
    info: '这里是用户的年龄',
    sorter: (a, b) => a.age - b.age,
    filters: ['小孩子', '大人'],
    onFilter: (value, record) => {
      if (value === '小孩子') {
        return record.age <= 22;
      }
      if (value === '大人') {
        return record.age > 22;
      }
      return false;
    },
  },
  {
    title: '体重',
    dataIndex: 'weight',
    key: 'weight',
    sorter: (a, b) => a.weight - b.weight,
    info: '这里是用户的体重',
    sortPriorityOrder: 2,
    filters: [60, 70],
    onFilter: (value, record) => {
      if (value === 60) {
        return record.weight === 60;
      }
      if (value === 70) {
        return record.weight === 70;
      }
      return false;
    },
  },
];

describe('Testing Table', () => {
  const getTable = () => <Table title="列表标题" dataSource={dataSource} columns={columns} pagination={false} />;

  it('should be stable', () => {
    const { asFragment } = render(getTable());
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const wrapper = mount(getTable());
      wrapper.setProps({ title: '列表标题设置' });
      wrapper.setProps({ pagination: true });
      wrapper.setProps({
        rowSelection: {
          onChange: (localSelectedRowKeys, selectRows) => {
            // eslint-disable-next-line no-console
            console.log(localSelectedRowKeys, selectRows);
          },
        },
      });
      wrapper.unmount();
    }).not.toThrow();
  });

  test('props prefixCls', () => {
    const wrapper = mount(getTable());
    wrapper.setProps({ prefixCls: 'custom-prefix' });
    expect(wrapper.exists('.custom-prefix-table')).toBe(true);
    expect(wrapper.exists('.gio-table')).toBe(false);
  });

  test('props className', () => {
    const wrapper = mount(getTable());
    wrapper.setProps({ className: 'custom-classname' });
    expect(wrapper.exists('.custom-classname')).toBe(true);
  });

  test('props title', () => {
    const wrapper = mount(getTable());
    expect(wrapper.exists('.gio-table-title')).toBe(true);
    expect(wrapper.find('.gio-table-title').at(0).text()).toBe('列表标题');
  });

  test('props pagination', () => {
    const onChange = jest.fn();
    const _dataSource = Array.from({ length: 100 }, (_, key) => ({ a: key, b: key, c: key, d: key }));
    const _columns = [
      {
        title: 'A',
        dataIndex: 'a',
        key: 'a',
        width: 200,
      },
      {
        title: 'B',
        dataIndex: 'b',
        key: 'b',
        width: 200,
      },
      {
        title: 'C',
        dataIndex: 'c',
        key: 'c',
        width: 200,
      },
      {
        title: 'D',
        dataIndex: 'd',
        key: 'd',
        width: 200,
      },
    ];
    const wrapper = mount(
      <Table title="列表标题" dataSource={_dataSource} columns={_columns} pagination onChange={onChange} />
    );
    expect(wrapper.exists('.gio-table-pagination')).toBe(true);
    wrapper.find('.gio-pagination-item').at(1).simulate('click');
    expect(onChange).toBeCalledWith({ current: 2, pageSize: 10 }, [], []);
  });

  it('should be render rightly', () => {
    const wrapper = mount(getTable());
    expect(wrapper.exists('.gio-table-title')).toBe(true);
    expect(wrapper.exists('.gio-table-column-title-info')).toBe(true);
    expect(wrapper.exists('.gio-table-column-filter')).toBe(true);
    expect(wrapper.exists('.gio-table-column-sorter')).toBe(true);
  });

  test('useHackOnRow hook', () => {
    const onClick = jest.fn();
    const onMouseUp = jest.fn();
    const onMouseDown = jest.fn();
    const wrapper = mount(getTable());
    wrapper.setProps({ hackRowEvent: true });
    wrapper.setProps({ onRow: () => ({ onClick, onMouseUp, onMouseDown }) });
    act(() => {
      wrapper.find('.gio-table-row').at(0).simulate('click');
      wrapper.find('.gio-table-row').at(0).simulate('mouseup');
      wrapper.find('.gio-table-row').at(0).simulate('mousedown');
    });
    expect(onClick).toBeCalled();
    expect(onMouseUp).toBeCalled();
    expect(onMouseDown).toBeCalled();
  });

  it('won`t trigger onRow click', () => {
    const onClick = jest.fn();
    const wrapper = mount(getTable());
    wrapper.setProps({ rowSelection: {}, onRow: () => ({ onClick }), onHeaderRow: () => ({ onClick }) });
    act(() => {
      wrapper.find('.gio-checkbox-input').at(0).simulate('click');
    });
    act(() => {
      wrapper.find('.gio-checkbox-input').at(1).simulate('click');
    });
    expect(onClick).not.toBeCalled();
    act(() => {
      wrapper.find('.gio-table-cell').at(1).simulate('click');
    });
    expect(onClick).toBeCalled();
  });

  it('will get Table Element instance', () => {
    let dom;
    mount(<Table ref={(_dom) => { dom = _dom }} />);
    expect(dom.className).toBe('gio-table-wrapper gio-table-showHover');
  })
});
