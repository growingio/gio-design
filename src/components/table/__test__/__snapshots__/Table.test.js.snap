// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Table should be stable 1`] = `
<DocumentFragment>
  <div
    class="gio-table-wrapper gio-table-showHover"
  >
    <div
      class="gio-loading-wrapper-loading"
    >
      <div
        class="gio-loading-container gio-loading-container-blur-white"
      >
        <div
          class="gio-table gio-table-layout-fixed"
        >
          <div
            class="gio-table-title"
          >
            列表标题
          </div>
          <div
            class="gio-table-container"
          >
            <div
              class="gio-table-content"
            >
              <table
                style="table-layout: fixed;"
              >
                <colgroup />
                <thead
                  class="gio-table-thead"
                >
                  <tr>
                    <th
                      class="gio-table-cell"
                    >
                      <div
                        class="gio-table-column-title"
                      >
                        姓名
                        <span
                          class="gio-table-column-title-info"
                        >
                          <span
                            aria-label="question-circle-outlined"
                            class="gio-icon"
                            role="img"
                          >
                            <svg
                              fill="currentColor"
                              height="16px"
                              viewBox="0 0 14 14"
                              width="16px"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <defs>
                                <style />
                              </defs>
                              <g>
                                <g>
                                  <path
                                    d="M7 13a6 6 0 114.24-1.76A6 6 0 017 13zM7 0a7 7 0 105 2.05A7 7 0 007 0zm.12 3a2.54 2.54 0 00-1.91.7 2.54 2.54 0 00-.69 1.69.24.24 0 00.05.14.2.2 0 00.15.06h.75a.21.21 0 00.2-.19A1.78 1.78 0 016 4.47 1.22 1.22 0 017 4a1.26 1.26 0 011 .33 1.25 1.25 0 01.32.9A1.27 1.27 0 018 6l-.19.21a5.43 5.43 0 00-1.26 1.4 2.15 2.15 0 00-.22 1 .21.21 0 00.19.19h.84a.15.15 0 00.15-.15 1.55 1.55 0 01.18-.73 1.81 1.81 0 01.46-.57 12.48 12.48 0 00.92-.86 2.1 2.1 0 00.42-1.34 2 2 0 00-.66-1.57A2.49 2.49 0 007.12 3zm-.2 6.43a.76.76 0 00-.57.22.73.73 0 00-.23.56.79.79 0 00.23.56.81.81 0 00.57.23.85.85 0 00.57-.22.78.78 0 00.24-.57.73.73 0 00-.23-.56.78.78 0 00-.58-.22z"
                                    fill="currentColor"
                                    id="question-circle-outlined_svg__question-circle"
                                  />
                                </g>
                              </g>
                            </svg>
                          </span>
                        </span>
                        <span
                          class="gio-table-column-sorter"
                        >
                          <span
                            class="gio-table-column-sorter-inner"
                          >
                            <button
                              class="gio-table-btn gio-table-column-sorter-inner-btn gio-table-btn-text gio-table-btn-icon-only"
                              type="button"
                            >
                              <span
                                aria-label="up-filled"
                                class="gio-icon gio-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="16px"
                                  viewBox="0 0 14 14"
                                  width="16px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M6.61 4.49a.5.5 0 01.78 0l3 3.7a.51.51 0 01.06.53A.5.5 0 0110 9H4a.5.5 0 01-.45-.28.51.51 0 01.06-.53z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="up-filled_svg__up-2"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                              <span
                                aria-label="down-filled"
                                class="gio-icon gio-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="16px"
                                  viewBox="0 0 14 14"
                                  width="16px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="down-filled_svg__down-2"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                            </button>
                          </span>
                        </span>
                        <span
                          class="gio-table-column-filter"
                        >
                          <span
                            class="gio-table-column-filter-inner"
                          >
                            <button
                              class="gio-btn gio-table-column-filter-inner-btn gio-btn-assist gio-btn-mini gio-btn-icon-only"
                              style="cursor: pointer;"
                              type="button"
                            >
                              <span
                                aria-label="filter-outlined"
                                class="gio-icon gio-table-column-filter-icon"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="12px"
                                  viewBox="0 0 14 14"
                                  width="12px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M6 6.46L5.56 6zM7.5 13a.51.51 0 00.5-.5V6.86a1 1 0 01.28-.69l4.16-4.32a.55.55 0 00.1-.55.51.51 0 00-.46-.3H1.92a.51.51 0 00-.46.3.53.53 0 00.1.55l4 4.15.16.17a1 1 0 01.28.69v5.64a.51.51 0 00.5.5zM9 13a1 1 0 01-1 1H6a1 1 0 01-1-1V7.28a1 1 0 00-.27-.69L.89 2.53A1.5 1.5 0 012 0h10a1.5 1.5 0 011.09 2.53L9.27 6.59a1 1 0 00-.27.69z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="filter-outlined_svg__filter"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                            </button>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      class="gio-table-cell"
                    >
                      <div
                        class="gio-table-column-title"
                      >
                        年龄
                        <span
                          class="gio-table-column-title-info"
                        >
                          <span
                            aria-label="question-circle-outlined"
                            class="gio-icon"
                            role="img"
                          >
                            <svg
                              fill="currentColor"
                              height="16px"
                              viewBox="0 0 14 14"
                              width="16px"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <defs>
                                <style />
                              </defs>
                              <g>
                                <g>
                                  <path
                                    d="M7 13a6 6 0 114.24-1.76A6 6 0 017 13zM7 0a7 7 0 105 2.05A7 7 0 007 0zm.12 3a2.54 2.54 0 00-1.91.7 2.54 2.54 0 00-.69 1.69.24.24 0 00.05.14.2.2 0 00.15.06h.75a.21.21 0 00.2-.19A1.78 1.78 0 016 4.47 1.22 1.22 0 017 4a1.26 1.26 0 011 .33 1.25 1.25 0 01.32.9A1.27 1.27 0 018 6l-.19.21a5.43 5.43 0 00-1.26 1.4 2.15 2.15 0 00-.22 1 .21.21 0 00.19.19h.84a.15.15 0 00.15-.15 1.55 1.55 0 01.18-.73 1.81 1.81 0 01.46-.57 12.48 12.48 0 00.92-.86 2.1 2.1 0 00.42-1.34 2 2 0 00-.66-1.57A2.49 2.49 0 007.12 3zm-.2 6.43a.76.76 0 00-.57.22.73.73 0 00-.23.56.79.79 0 00.23.56.81.81 0 00.57.23.85.85 0 00.57-.22.78.78 0 00.24-.57.73.73 0 00-.23-.56.78.78 0 00-.58-.22z"
                                    fill="currentColor"
                                    id="question-circle-outlined_svg__question-circle"
                                  />
                                </g>
                              </g>
                            </svg>
                          </span>
                        </span>
                        <span
                          class="gio-table-column-sorter"
                        >
                          <span
                            class="gio-table-column-sorter-inner"
                          >
                            <button
                              class="gio-table-btn gio-table-column-sorter-inner-btn gio-table-btn-text gio-table-btn-icon-only"
                              type="button"
                            >
                              <span
                                aria-label="up-filled"
                                class="gio-icon gio-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="16px"
                                  viewBox="0 0 14 14"
                                  width="16px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M6.61 4.49a.5.5 0 01.78 0l3 3.7a.51.51 0 01.06.53A.5.5 0 0110 9H4a.5.5 0 01-.45-.28.51.51 0 01.06-.53z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="up-filled_svg__up-2"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                              <span
                                aria-label="down-filled"
                                class="gio-icon gio-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="16px"
                                  viewBox="0 0 14 14"
                                  width="16px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="down-filled_svg__down-2"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                            </button>
                          </span>
                        </span>
                        <span
                          class="gio-table-column-filter"
                        >
                          <span
                            class="gio-table-column-filter-inner"
                          >
                            <button
                              class="gio-btn gio-table-column-filter-inner-btn gio-btn-assist gio-btn-mini gio-btn-icon-only"
                              style="cursor: pointer;"
                              type="button"
                            >
                              <span
                                aria-label="filter-outlined"
                                class="gio-icon gio-table-column-filter-icon"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="12px"
                                  viewBox="0 0 14 14"
                                  width="12px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M6 6.46L5.56 6zM7.5 13a.51.51 0 00.5-.5V6.86a1 1 0 01.28-.69l4.16-4.32a.55.55 0 00.1-.55.51.51 0 00-.46-.3H1.92a.51.51 0 00-.46.3.53.53 0 00.1.55l4 4.15.16.17a1 1 0 01.28.69v5.64a.51.51 0 00.5.5zM9 13a1 1 0 01-1 1H6a1 1 0 01-1-1V7.28a1 1 0 00-.27-.69L.89 2.53A1.5 1.5 0 012 0h10a1.5 1.5 0 011.09 2.53L9.27 6.59a1 1 0 00-.27.69z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="filter-outlined_svg__filter"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                            </button>
                          </span>
                        </span>
                      </div>
                    </th>
                    <th
                      class="gio-table-cell"
                    >
                      <div
                        class="gio-table-column-title"
                      >
                        体重
                        <span
                          class="gio-table-column-title-info"
                        >
                          <span
                            aria-label="question-circle-outlined"
                            class="gio-icon"
                            role="img"
                          >
                            <svg
                              fill="currentColor"
                              height="16px"
                              viewBox="0 0 14 14"
                              width="16px"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <defs>
                                <style />
                              </defs>
                              <g>
                                <g>
                                  <path
                                    d="M7 13a6 6 0 114.24-1.76A6 6 0 017 13zM7 0a7 7 0 105 2.05A7 7 0 007 0zm.12 3a2.54 2.54 0 00-1.91.7 2.54 2.54 0 00-.69 1.69.24.24 0 00.05.14.2.2 0 00.15.06h.75a.21.21 0 00.2-.19A1.78 1.78 0 016 4.47 1.22 1.22 0 017 4a1.26 1.26 0 011 .33 1.25 1.25 0 01.32.9A1.27 1.27 0 018 6l-.19.21a5.43 5.43 0 00-1.26 1.4 2.15 2.15 0 00-.22 1 .21.21 0 00.19.19h.84a.15.15 0 00.15-.15 1.55 1.55 0 01.18-.73 1.81 1.81 0 01.46-.57 12.48 12.48 0 00.92-.86 2.1 2.1 0 00.42-1.34 2 2 0 00-.66-1.57A2.49 2.49 0 007.12 3zm-.2 6.43a.76.76 0 00-.57.22.73.73 0 00-.23.56.79.79 0 00.23.56.81.81 0 00.57.23.85.85 0 00.57-.22.78.78 0 00.24-.57.73.73 0 00-.23-.56.78.78 0 00-.58-.22z"
                                    fill="currentColor"
                                    id="question-circle-outlined_svg__question-circle"
                                  />
                                </g>
                              </g>
                            </svg>
                          </span>
                        </span>
                        <span
                          class="gio-table-column-sorter"
                        >
                          <span
                            class="gio-table-column-sorter-inner"
                          >
                            <button
                              class="gio-table-btn gio-table-column-sorter-inner-btn gio-table-btn-text gio-table-btn-icon-only"
                              type="button"
                            >
                              <span
                                aria-label="up-filled"
                                class="gio-icon gio-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="16px"
                                  viewBox="0 0 14 14"
                                  width="16px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M6.61 4.49a.5.5 0 01.78 0l3 3.7a.51.51 0 01.06.53A.5.5 0 0110 9H4a.5.5 0 01-.45-.28.51.51 0 01.06-.53z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="up-filled_svg__up-2"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                              <span
                                aria-label="down-filled"
                                class="gio-icon gio-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="16px"
                                  viewBox="0 0 14 14"
                                  width="16px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M7.39 9.51a.5.5 0 01-.78 0l-3-3.7a.51.51 0 01-.06-.53A.5.5 0 014 5h6a.5.5 0 01.45.28.51.51 0 01-.06.53z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="down-filled_svg__down-2"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                            </button>
                          </span>
                        </span>
                        <span
                          class="gio-table-column-filter"
                        >
                          <span
                            class="gio-table-column-filter-inner"
                          >
                            <button
                              class="gio-btn gio-table-column-filter-inner-btn gio-btn-assist gio-btn-mini gio-btn-icon-only"
                              style="cursor: pointer;"
                              type="button"
                            >
                              <span
                                aria-label="filter-outlined"
                                class="gio-icon gio-table-column-filter-icon"
                                role="img"
                              >
                                <svg
                                  fill="currentColor"
                                  height="12px"
                                  viewBox="0 0 14 14"
                                  width="12px"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <defs>
                                    <style />
                                  </defs>
                                  <g>
                                    <g>
                                      <path
                                        d="M6 6.46L5.56 6zM7.5 13a.51.51 0 00.5-.5V6.86a1 1 0 01.28-.69l4.16-4.32a.55.55 0 00.1-.55.51.51 0 00-.46-.3H1.92a.51.51 0 00-.46.3.53.53 0 00.1.55l4 4.15.16.17a1 1 0 01.28.69v5.64a.51.51 0 00.5.5zM9 13a1 1 0 01-1 1H6a1 1 0 01-1-1V7.28a1 1 0 00-.27-.69L.89 2.53A1.5 1.5 0 012 0h10a1.5 1.5 0 011.09 2.53L9.27 6.59a1 1 0 00-.27.69z"
                                        fill="currentColor"
                                        fill-rule="evenodd"
                                        id="filter-outlined_svg__filter"
                                      />
                                    </g>
                                  </g>
                                </svg>
                              </span>
                            </button>
                          </span>
                        </span>
                      </div>
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="gio-table-tbody"
                >
                  <tr
                    class="gio-table-row gio-table-row-level-0"
                    data-row-key="1"
                  >
                    <td
                      class="gio-table-cell"
                    >
                      胡彦斌
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      32
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      70
                    </td>
                  </tr>
                  <tr
                    class="gio-table-row gio-table-row-level-0"
                    data-row-key="2"
                  >
                    <td
                      class="gio-table-cell"
                    >
                      胡彦祖
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      42
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      60
                    </td>
                  </tr>
                  <tr
                    class="gio-table-row gio-table-row-level-0"
                    data-row-key="3"
                  >
                    <td
                      class="gio-table-cell"
                    >
                      航航
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      18
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      70
                    </td>
                  </tr>
                  <tr
                    class="gio-table-row gio-table-row-level-0"
                    data-row-key="4"
                  >
                    <td
                      class="gio-table-cell"
                    >
                      屁屁
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      22
                    </td>
                    <td
                      class="gio-table-cell"
                    >
                      60
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
